lib/ner/gpt3.ts

```diff
@@ -0,0 +1,306 @@
+import fs from 'fs';
+import * as Tp from 'thingpedia';
+import { Entity, Linker } from './base';
+import WikidataUtils from '../utils/wikidata';
+import Cache from '../utils/cache';
+
+interface GPT3EntityLinkerOptions {
+    ner_cache : string,
+    raw_data ?: string
+}
+
+export class GPT3Linker extends Linker {
+    private _wikidata : WikidataUtils;
+    private _url : string;
+    private _cache : Cache;
+    private _rawData : Record<string, string>;
+    // private _timeOfInceptionProperties : string[];
+
+    constructor(wikidata : WikidataUtils, options : GPT3EntityLinkerOptions) {
+        super();
+        this._wikidata = wikidata;
+        this._url = 'https://wikidata.openai.azure.com/openai/deployments/text/completions?api-version=2022-12-01';
+        this._cache = new Cache(options.ner_cache);
+        this._rawData = {};
+        // this._timeOfInceptionProperties = ["P569", "P580", "P577", "P571", "P585", "P575"];
+        if (options.raw_data) {
+            for (const ex of JSON.parse(fs.readFileSync(options.raw_data, 'utf-8')).questions)
+                this._rawData[ex.id] = ex.question[0].string;
+        }
+    }
+
+    async run(id : string, utterance : string) {
+        if (id in this._rawData)
+            utterance = this._rawData[id];
+        const cache = await this._cache.get(utterance);
+        if (cache)
+            return JSON.parse(cache);
+        const entities : Entity[] = [];
+
+        if (!process.env.OPENAI_API_KEY)
+            throw new Error('Missing environment variable OPENAI_API_KEY');
+
+        const prompt = this._prompt(utterance);
+
+        const raw = await Tp.Helpers.Http.post(this._url, JSON.stringify({ prompt, max_tokens: 500, temperature: 0 }), {
+             dataContentType: 'application/json',
+             extraHeaders: { 'api-key': process.env.OPENAI_API_KEY as string }
+        });
+
+        const res = JSON.parse(raw);
+        if (res.choices.length > 0 && res.choices[0].text.length > 0) {
+            const entitiesFound = res.choices[0].text.trim().split('\n').slice(1);
+            for (const entity of entitiesFound) {
+                if (!entity)
+                    continue;
+                const [name, toi, propertiesReturnedUncleaned] = entity.split('; ');
+                let propertiesReturned = '';
+                if (propertiesReturnedUncleaned) {
+                    if (propertiesReturnedUncleaned.trim().substring(propertiesReturnedUncleaned.length - 1) === '#')
```

Comment:
So the response sometimes has `#` and sometimes doesn't? 
btw, you can do `.trim().endsWith('#')`, or just `.trim().replace('#', '')`